virttest/cpu.py

```diff
@@ -1258,3 +1261,110 @@ def get_qemu_best_cpu_model(params):
             return host_cpu_model
     # If no host cpu model can be found on qemu_cpu_models, choose the default
     return params.get("default_cpu_model", None)
+
+
+def get_qemu_best_cpu_info(params):
+    """
+    Get the best cpu model and features available for qemu
+    """
+    cpu_info = {}
+    cpu_model = get_qemu_best_cpu_model(params)
+    # Expand if needed
+    cpu_flags = ""
+    cpu_info["model"] = cpu_model
+    cpu_info["flags"] = cpu_flags
+    return cpu_info
+
+
+def parse_qemu_cpu_flags(cpu_flags_str):
+    """
+    Convert cpu flags string to a dict
+    """
+    flags_dict = {}
+    if cpu_flags_str:
+        flags_list = cpu_flags_str.lstrip(",").split(",")
+        for item in flags_list:
+            if item.startswith("+"):
+                flags_dict[item[1:]] = "on"
+            elif item.startswith("-"):
+                flags_dict[item[1:]] = "off"
+            elif item.find("=") != -1:
+                item = item.split("=")
+                flags_dict[item[0]] = item[1]
+            else:
+                flags_dict[item] = None
+    return flags_dict
+
+
+def get_cpu_info_from_virsh_qemu_cli(name):
+    """
+    Get cpu info from qemu cmdline generated by virsh
+    """
+    cmd = "ps aux | grep qemu | grep -e '\s-name\s\+guest=%s\(\,\|\s\)'" % name
+    output = process.getoutput(cmd, shell=True)
+    cpu_line = re.findall("-cpu\s+(\S*)\s+", output)[0]
+    cpu_cmd = cpu_line.split(',', 1)
+    cpu_info = {}
+    cpu_info["model"] = cpu_cmd[0]
+    cpu_info["flags"] = "" if len(cpu_cmd) < 2 else cpu_cmd[1]
+    return cpu_info
+
+
+def get_cpu_info_from_virsh(params):
+    """
+    Try to get cpu model and features from virsh with 'host-model'
+    """
+    vm_arch = params["vm_arch_name"]
+    if vm_arch not in ['x86_64', 'i686']:
+        raise NotImplementedError("Arch '%s' is not supported" % vm_arch)
+    machine = params.get('machine_type')
+    name = uuid.uuid4().hex
+    try:
+        path.find_command("virsh")
+    except path.CmdNotFoundError:
+        logging.warning("Virsh executable not set or found on path")
+        return
+
+    xml = """
+            <domain type='kvm'>
+                <name>%s</name>
+                <memory>1</memory>
+                <os>
+                    <type arch='%s' machine='%s'>hvm</type>
+                </os>
+                <cpu mode='host-model'/>
+            </domain>
+          """ % (name, vm_arch, machine)
+    xml_file = os.path.join(data_dir.get_tmp_dir(), "temp_xml_for_cpu")
+    with open(xml_file, "w") as f:
+        f.write(xml)
+    try:
+        logging.info("Get cpu model and features from virsh")
+        virsh.define(xml_file)
```

Comment:
Hi @huangyum 
Recently, we hit a hanging error when the virsh is defining xml since we didn't add the timeout for this process which will result in the job always hanging.
The gdb info:
```
Traceback (most recent call first):
  <built-in method waitpid of module object at remote 0x7f59bc9a42c0>
  File "/usr/lib64/python3.9/subprocess.py", line 1891, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
  File "/usr/lib64/python3.9/subprocess.py", line 1933, in _wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib64/python3.9/subprocess.py", line 1189, in wait
    return self._wait(timeout=timeout)
  File "/home/kar/workspace/avocado/avocado/utils/process.py", line 856, in wait
    rc = self._popen.wait()
  File "/home/kar/workspace/avocado/avocado/utils/process.py", line 942, in run
    self.wait(timeout, sig)
  File "/home/kar/workspace/avocado/avocado/utils/process.py", line 1009, in run
    cmd_result = sp.run(timeout=timeout)
  File "/home/kar/workspace/avocado-vt/virttest/virsh.py", line 819, in command
    ret = process.run(cmd, timeout=timeout, verbose=debug,
  File "/home/kar/workspace/avocado-vt/virttest/virsh.py", line 1580, in define
    return command(cmd, **dargs)
  File "/home/kar/workspace/avocado-vt/virttest/cpu.py", line 1351, in get_cpu_info_from_virsh
    virsh.define(xml_file)
--Type <RET> for more, q to quit, c to continue without paging--
  File "/home/kar/workspace/avocado-vt/virttest/env_process.py", line 1351, in preprocess
    cpu_info = policy_map[policy](params)
  File "/home/kar/workspace/avocado-vt/virttest/error_context.py", line 135, in new_fn
    return fn(*args, **kwargs)
  File "/home/kar/workspace/avocado-vt/avocado_vt/test.py", line 252, in _runTest
    params = env_process.preprocess(self, params, env)
....
```

CC @qingwangrh

Reply:
Hi @YongxueHong , just checked all the code calling virsh.define, none of them add timeout, wondering why the issue is never hit before. Anyway, I agree 'timeout' should be set for any process in case of process hanging and never return. Adding timeout when invoking virsh.define looks like a workaround to me, maybe we could move the `if` block in https://github.com/avocado-framework/avocado-vt/blob/master/virttest/virsh.py#L792C7-L792C7 out of `if session:` in L788. What do you think?

Reply:
Hi @huangyum 
For your solution, we need to get the session id of an existing `VirshSession` in advance, 
and the way does not support `readonly` mode if we want to use `virsh.define` with `readonly` mode in the future.
and we can not support other options such as `virsh_opt`, `uri`
https://github.com/avocado-framework/avocado-vt/blob/fe0abdc708399b9f2b9e9321b2395b11a48b9047/virttest/virsh.py#L788-L795

And the timeout is default 60 if we don't pass the timeout here.
In my point of view, actually, this solution is the same as virsh.define(xml_file, None, timeout=60)

So I prefer to specify the argument timeout when calling `virsh.define`. Please let me know your opinion.

And we should consider adding the timeout for other places where call the `virsh` method without the limitation of the timeout, like `virsh.start(name)`, `virsh.destroy(name, ignore_status=True)` and `virsh.undefine(name, ignore_status=True)`.

Reply:
> Hi @huangyum For your solution, we need to get the session id of an existing `VirshSession` in advance, and the way does not support `readonly` mode if we want to use `virsh.define` with `readonly` mode in the future. and we can not support other options such as `virsh_opt`, `uri`
> 
> https://github.com/avocado-framework/avocado-vt/blob/fe0abdc708399b9f2b9e9321b2395b11a48b9047/virttest/virsh.py#L788-L795
> 
> And the timeout is default 60 if we don't pass the timeout here. In my point of view, actually, this solution is the same as virsh.define(xml_file, None, timeout=60)
> 
> So I prefer to specify the argument timeout when calling `virsh.define`. Please let me know your opinion.
> 
> And we should consider adding the timeout for other places where call the `virsh` method without the limitation of the timeout, like `virsh.start(name)`, `virsh.destroy(name, ignore_status=True)` and `virsh.undefine(name, ignore_status=True)`.

Hi @huangyum 
What's your opinion?

Reply:
Agree, sent out the patch https://github.com/avocado-framework/avocado-vt/pull/3754. Thanks.